namer/namer.cc

```diff
@@ -15,388 +18,794 @@
+            existing = ctx.state.enterClassSymbol(core::Loc(ctx.file, loc), scope, name);
+            existing.data(ctx)->singletonClass(ctx); // force singleton class into existance
         }
 
-        ast::InsSeq::STATS_store retSeqs;
-        auto loc = klass->declLoc;
-        retSeqs.emplace_back(std::move(klass));
-        for (auto &stat : ideSeqs) {
-            retSeqs.emplace_back(std::move(stat));
+        return existing;
+    }
+
+    void defineArg(core::MutableContext ctx, int pos, const ast::ParsedArg &parsedArg) {
+        if (pos < ctx.owner.data(ctx)->arguments().size()) {
+            // TODO: check that flags match;
+            ctx.owner.data(ctx)->arguments()[pos].loc = core::Loc(ctx.file, parsedArg.loc);
+            return;
         }
-        return ast::MK::InsSeq(loc.offsets(), std::move(retSeqs), ast::MK::EmptyTree());
+
+        core::NameRef name;
+        if (parsedArg.flags.isKeyword) {
+            name = parsedArg.local._name;
+        } else if (parsedArg.flags.isBlock) {
+            name = core::Names::blkArg();
+        } else {
+            name = ctx.state.freshNameUnique(core::UniqueNameKind::PositionalArg, core::Names::arg(), pos + 1);
+        }
+        // we know right now that pos >= arguments().size() because otherwise we would have hit the early return at the
+        // beginning of this method
+        auto &argInfo = ctx.state.enterMethodArgumentSymbol(core::Loc(ctx.file, parsedArg.loc), ctx.owner, name);
+        // if enterMethodArgumentSymbol did not emplace a new argument into the list, then it means it's reusing an
+        // existing one, which means we've seen a repeated kwarg (as it treats identically named kwargs as
+        // identical). We know that we need to match the arity of the function as written, so if we don't have as many
+        // arguments as we expect, clone the one we got back from enterMethodArgumentSymbol in the position we expect
+        if (ctx.owner.dataAllowingNone(ctx)->arguments().size() == pos) {
+            auto argCopy = argInfo.deepCopy();
+            argCopy.name = ctx.state.freshNameUnique(core::UniqueNameKind::MangledKeywordArg, argInfo.name, pos + 1);
+            ctx.owner.dataAllowingNone(ctx)->arguments().emplace_back(move(argCopy));
+            return;
+        }
+        // at this point, we should have at least pos + 1 arguments, and arguments[pos] should be the thing we got back
+        // from enterMethodArgumentSymbol
+        ENFORCE(ctx.owner.data(ctx)->arguments().size() >= pos + 1);
+
+        argInfo.flags = parsedArg.flags;
     }
 
-    ast::MethodDef::ARGS_store fillInArgs(core::MutableContext ctx, ast::MethodDef::ARGS_store &oldArgs,
-                                          vector<ast::ParsedArg> parsedArgs) {
-        ENFORCE(oldArgs.size() == parsedArgs.size());
-        ast::MethodDef::ARGS_store args;
+    void defineArgs(core::MutableContext ctx, const vector<ast::ParsedArg> &parsedArgs) {
```

Comment:
let's have this one take SymbolData&. All callsites that call it have access to it